carbon-design-system/carbon-addons-iot-react
# fix(stateful table): enable empty string filtering in combo-box

Closes #3529

**Summary**

- Add special case filtering by empty string

**Change List (commits, features, bugs, etc)**

- Modify filter handlers in `FilterHeaderRow` to enable filtering by empty string
- Add unit test
- Update Storybook to reproduce the issue

**Acceptance Test (how to verify the PR)**

- __Single-select filter__
- Go to [this story](https://deploy-preview-3552--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table--with-filtering)
- In the Knobs section enable `Show empty string option in Select filter`
- Reset current filters
- Select `Unspecified` in `Select` combo-box
- Verify that table has been filtered according to empty value in `Select` column

<br/>

- __Multi-select filter__
- Same story
- Enable `Show empty string option in Select filter`
- Enable `Allow multi-select in Select filter`
- Reset current filters
- Select `Unspecified` in `Select` combo-box
- Verify that table has been filtered according to empty value in `Select` column
- Select second option in the multiselect (for example `option-A`)
- Verify that table has been filtered according to empty value in `Select` column

<br/>

- __Advanced filter__
- Same story
- Enable `Show empty string option in Select filter`
- Enable `Allow multi-select in Select filter`
- Enable `Enable advanced filters`
- Reset auto-applied advanced filters
- Click on the filter icon
- In the Simple filters tab:
- Select `Unspecified` in `Select` combo-box
- Click on `Apply filters`
- Verify that table has been filtered according to empty value in `Select` column
- Select second option in the multiselect (for example `option-A`)
- Click on `Apply filters`
- Verify that table has been filtered according to empty value in `Select` column

**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here



**Things to look for during review**
